src/components/Context.js

```diff
@@ -5,45 +5,49 @@ import _ from "lodash";
 
 const ProductContext = React.createContext();
 
-const REMOVE_PROMOTION = "REMOVE_PROMOTION";
 const SET_CART = "SET_CART";
 const CLEAR_CART = "CLEAR_CART";
+const SET_COUPON = "SET_COUPON";
 
 const reducer = (action) => (state, props) => {
-  switch (action.type) {
-    case SET_CART:
-      let discountedAmount = 0;
-      let tempTotal = 0;
-      const cartTotal = state.cart.map((item) => (tempTotal += item.total));
-      let cartTotalAfterPromotion = cartTotal;
-
-      if (!_.isEmpty(state.appliedVoucher)) {
-        if (state.appliedVoucher.discount.type === "PERCENT") {
-          const discountAmount = state.appliedVoucher.discount.percent_off;
-          cartTotalAfterPromotion =
-            cartTotal - cartTotal * (discountAmount / 100);
-          discountedAmount = cartTotal * (discountAmount / 100);
-          console.log(discountAmount);
-        } else if (state.appliedVoucher.discount.type === "AMOUNT") {
-          const discountAmount = state.appliedVoucher.discount.amount;
-          cartTotalAfterPromotion = cartTotal - discountAmount;
-          discountedAmount = discountAmount;
-        }
+  const calc = (cartItems, voucher) => {
+    const cartTotal = cartItems.reduce((sum, item) => sum + item.total, 0);
+    let discountedAmount = 0;
+    let cartTotalAfterPromotion = cartTotal;
+
+    if (!_.isEmpty(voucher)) {
+      if (voucher.discount.type === "PERCENT") {
+        const discountAmount = voucher.discount.percent_off;
+        cartTotalAfterPromotion =
+          cartTotal - cartTotal * (discountAmount / 100);
+        discountedAmount = cartTotal * (discountAmount / 100);
+        console.log(discountAmount);
+      } else if (voucher.discount.type === "AMOUNT") {
+        const discountAmount = voucher.discount.amount;
+        cartTotalAfterPromotion = cartTotal - discountAmount;
+        discountedAmount = discountAmount;
       }
+    }
+    return {
+      cart: cartItems,
+      discountedAmount,
+      cartTotal,
+      cartTotalAfterPromotion,
+      appliedVoucher: voucher,
+    }
+  };
 
-      return {
-        cartTotalAfterPromotion: cartTotalAfterPromotion,
-        discountedAmount: discountedAmount,
-        cartTotal: cartTotal,
-      };
-
-    case REMOVE_PROMOTION:
-      return {
-        appliedVoucher: {},
-        cartTotalAfterPromotion: state.cartTotal,
-      };
-
+  switch (action.type) {
+    case SET_COUPON:
+      const calculatedCart = calc(state.cart, action.appliedVoucher);
```

Comment:
for SET_COUPON action we expect that the voucher has changed, but the cart items did not .
so we take `cart` from the state, but use the `appliedVoucher` fromthe action payload